### PR TITLE
Dummies multi backend

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

We now have several objects that requires two of the backends and not just one. This PR adds support for dummies that require two backends.

To achieve that the various `requires_xxx` function used internally are refactored in one `requires_backends` function that takes one or a list of backends.

### CODE REVIEW

File: src/transformers/__init__.py [hunk 1]
```diff
@@ -339,30 +339,42 @@
...
-
 else:
     from .utils import dummy_tokenizers_objects
 
     _import_structure["utils.dummy_tokenizers_objects"] = [
         name for name in dir(dummy_tokenizers_objects) if not name.startswith("_")
     ]
 
+if is_sentencepiece_available() and is_tokenizers_available():
+    _import_structure["convert_slow_tokenizer"] = ["SLOW_TO_FAST_CONVERTERS", "convert_slow_tokenizer"]
```
Comment (sgugger) [inline src/transformers/__init__.py:350]: Same import as L344 but in a proper test.

File: src/transformers/models/rag/retrieval_rag.py [hunk 2]
```diff
@@ -372,8 +365,7 @@ class RagRetriever:
 
     def __init__(self, config, question_encoder_tokenizer, generator_tokenizer, index=None, init_retrieval=True):
         self._init_retrieval = init_retrieval
-        requires_datasets(self)
-        requires_faiss(self)
+        requires_backends(self, ["datasets", "faiss"])
```
Comment (sgugger) [inline src/transformers/models/rag/retrieval_rag.py:368]: This kind of refactor is nice and will give one clear error message containing all required dependencies.

File: utils/check_dummies.py [hunk 3]
```diff
@@ -156,8 +160,11 @@ def check_dummies(overwrite=False):
 
     actual_dummies = {}
     for backend, file_path in dummy_file_paths.items():
-        with open(file_path, "r", encoding="utf-8", newline="\n") as f:
-            actual_dummies[backend] = f.read()
+        if os.path.isfile(file_path):
+            with open(file_path, "r", encoding="utf-8", newline="\n") as f:
+                actual_dummies[backend] = f.read()
+        else:
+            actual_dummies[backend] = ""
```
Comment (sgugger) [inline utils/check_dummies.py:167]: This removes the error that appeared when running `make fix-copies` and the corresponding dummy files did not exist yet.

File: src/transformers/models/rag/retrieval_rag.py (hunk 2 above)
Comment (LysandreJik) [reply src/transformers/models/rag/retrieval_rag.py:368]: This is very nice indeed

**Review by LysandreJik — APPROVED** [review]
This is very nice! Thanks a lot for taking care of it.

**Pull request merged**